internal/pkg/earpointer/analysis.go

```diff
@@ -0,0 +1,635 @@
+	case token.MUL:
+		{ // v = *addr
+			// The LHS is a register; the RHS can be a (memory) address, a global
+			// variable, or a free variable.
+			// t1 = *addr is implemented as "addr[directPointToField] = t1",
+			// i.e., address addr points to value t1.
+			p.processAddressToValue(op.X, op /*unifyByValue=*/, false)
+		}
+	}
+}
+
+func (p *transformer) visitBinOp(op *ssa.BinOp) {
+	p.unifyLocals(op, op.X)
+	p.unifyLocals(op, op.Y)
+}
+
+func (p *transformer) visitTypeAssert(assert *ssa.TypeAssert) {
+	if !assert.CommaOk {
+		// t1 = typeassert t0.(*int)
+		p.unifyLocals(assert, assert.X)
+	}
+	// The comma case will be processed by the subsequent extract instruction.
+}
+
+func (p *transformer) visitExtract(extract *ssa.Extract) {
+	switch base := extract.Tuple.(type) {
+	case *ssa.TypeAssert:
+		// t1 = typeassert,ok t0.(*int)
+		// t2 = extract t1 0
+		if base.CommaOk && extract.Index == 0 {
+			p.unifyLocals(extract, base.X)
+		}
+	case *ssa.Next:
+		// t1 = range t0
+		// t2 = next t1   // return (ok bool, k Key, v Value)
+		// t3 = extract t2 ?
+		if rng, ok := base.Iter.(*ssa.Range); ok {
+			if mayShareObject(extract) && extract.Index == 2 { // extract the value v
+				// The value from the map flows into the field: m[*] = v.
+				p.processHeapAccess(extract, rng.X, anyIndexField,
+					/*unify_by_value=*/ false)
+			}
+		}
+	case *ssa.Lookup:
+		// t2 = t0[t1],ok
+		// t3 = extract t2 ?
+		if base.CommaOk && extract.Index == 0 {
+			p.processHeapAccess(extract, base.X, Field{Name: "0"},
+				/*unify_by_value=*/ false)
+		}
+	default: // Other cases: model "r1 = extract r0 #n" as "r0["n"] = r1".
+		field := Field{Name: strconv.Itoa(extract.Index)}
+		p.processHeapAccess(extract, base, field,
+			/*unify_by_value=*/ false)
+	}
+}
+
+func (p *transformer) visitNext(next *ssa.Next) {
+	// The next instruction will be processed by the subsequent extract
+	// instruction.
+}
+
+func (p *transformer) visitSlice(slice *ssa.Slice) {
+	p.unifyLocals(slice, slice.X)
+}
+
+func (p *transformer) visitMakeInterface(makeInterface *ssa.MakeInterface) {
+	if mayShareObject(makeInterface.X) {
+		p.unifyLocals(makeInterface, makeInterface.X)
+	}
+}
+
+func (p *transformer) visitSend(send *ssa.Send) {
+	p.processHeapAccess(send.X, send.Chan,
+		anyIndexField /*unify_by_value=*/, false)
+}
+
+func (p *transformer) visitRange(rng *ssa.Range) {
+	// t1 = range t0: unify t0 and t1.
+	p.unifyLocals(rng, rng.X)
+}
+
+func (p *transformer) VisitSelect() {
+	// TODO: to be implemented.
+}
+
+func (p *transformer) visitCall(call ssa.CallCommon, instr ssa.Instruction) {
+	// TODO: to be added.
+}
+
+// Process a load/store using an index (which can be constant).
+func (p *transformer) processMemoryIndexAccess(data ssa.Value, base ssa.Value, index ssa.Value, unifyByValue bool) {
+	// If the index is a constant, its string name is used as the field name;
+	// otherwise the predefined field name is used.
+	var field Field
+	if c, ok := index.(*ssa.Const); ok {
```

Comment:
It tries to distinguish the field accesses. For example, 

type T struct { x *int, y *int }
var t T

the analysis knows that t.x is different from t.y.  
Array and Slice accesses are also modeled using fields. However in some cases, e.g. when symbolic indices is used for arrays, e.g. arr[1] and arr[x], the analysis don't know about whether x can or must equal to 1, so approximation is used.

Reply:
It can also lead to underapproximation for maps, e.g.:
```go
	func f(i int, xs map[int]*int) {
		xs[i] = &i
		xs[0] = &i
	}
```
... yields this state:
```
{f.t0}: [], {f.xs}: [0->f.t0, AnyField->f.t0]
```
... and similarly for map `Lookup`s, e.g. `_, _ = xs[0], xs[i]`.
	}

Reply:
Right. Currently the analysis doesn't resolve symbolic indices like the i in this example. So xs[i] is represented by AnyField-> t0. It also maintains another field "0" since it knows that that field points to t0. An optimization is to combine these two fields into AnyField-> t0 and remove field "0". I keep both because I will leave it to the user to decide how to use them. For example, 

{xs}: [0->t0, AnyField->t1]

indicates that xs[0] can be {t0, t1}, and another index i != 0 can be {t1}. The user knows that xs[1] won't be t0, so we have a little bit better accuracy here. Otherwise, if we combine them to be  {xs}: [AnyField->{t0, t1}], then it indicates that xs[1] can point to t0, which is an over approximation. 
Of course, this representation will require the user to do extract work when using the heap structure.